HTTP/2 test intermittent failure

We've seen multiple, intermittent failures of the following test in the pipeline:

io.helidon.tests.apps.bookstore.se.Http2SslTest

Test uses somewhat old okhttp client. Output is:

```
Error:  Tests run: 4, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 1.001 s <<< FAILURE! - in io.helidon.tests.apps.bookstore.se.Http2SslTest
Error:  io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2SslPostFirst  Time elapsed: 0.054 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <200>
     but: was <409>
	at io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2SslPostFirst(Http2SslTest.java:77)

Error:  io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2SslCompression  Time elapsed: 0.104 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <200>
     but: was <409>
	at io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorld(Http2SslTest.java:119)
	at io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2SslCompression(Http2SslTest.java:68)

Error:  io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2Ssl  Time elapsed: 0.111 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <200>
     but: was <409>
	at io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorld(Http2SslTest.java:119)
	at io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2Ssl(Http2SslTest.java:63)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    Http2SslTest.testHelloWorldHttp2Ssl:63->testHelloWorld:119 
Expected: is <200>
     but: was <409>
Error:    Http2SslTest.testHelloWorldHttp2SslCompression:68->testHelloWorld:119 
Expected: is <200>
     but: was <409>
Error:    Http2SslTest.testHelloWorldHttp2SslPostFirst:77 
Expected: is <200>
     but: was <409>
Error:    SslTest.testHelloWorldSsl:64 
Expected: is <200>
     but: was <409>
[INFO] 
Error:  Tests run: 6, Failures: 4, Errors: 0, Skipped: 0
```

With some test tweaks (the test incorrectly uses HTTP/2 in some cases), the following output was produced:

```
2023.06.23 18:17:46 INFO io.helidon.nima.webserver.ServerListener Thread[#44,server-@default-listener,5,main]: [0x01eb6e1c] @default socket closed.
2023.06.23 18:17:46 INFO io.helidon.nima.webserver.LoomServer Thread[#1,main,5,main]: Níma server stopped all channels.
Error:  Tests run: 4, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.928 s <<< FAILURE! - in io.helidon.tests.apps.bookstore.se.Http2SslTest
Error:  io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2SslCompression  Time elapsed: 0.007 s  <<< ERROR!
java.io.IOException: TYPE_GOAWAY unexpected error code: 5

Error:  io.helidon.tests.apps.bookstore.se.Http2SslTest.testHelloWorldHttp2Ssl  Time elapsed: 0.136 s  <<< ERROR!
java.io.IOException: TYPE_GOAWAY unexpected error code: 5

[INFO] 
[INFO] Results:
[INFO] 
Error:  Errors: 
Error:    Http2SslTest.testHelloWorldHttp2Ssl » IO TYPE_GOAWAY unexpected error code: 5
Error:    Http2SslTest.testHelloWorldHttp2SslCompression » IO TYPE_GOAWAY unexpected err...
Error:    SslTest.testHelloWorldSsl » IO TYPE_GOAWAY unexpected error code: 5
[INFO] 
```

--- a/tests/apps/bookstore/bookstore-se/pom.xml
+++ b/tests/apps/bookstore/bookstore-se/pom.xml
@@ -92,6 +92,12 @@
             <version>${project.version}</version>
         </dependency>
         <dependency>
+            <groupId>io.helidon.nima.testing.junit5
+            </groupId>
+            <artifactId>helidon-nima-testing-junit5-webserver</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
             <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter-api</artifactId>
             <scope>test</scope>
@@ -99,12 +105,6 @@
         <dependency>
             <groupId>org.hamcrest</groupId>
             <artifactId>hamcrest-all</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>com.squareup.okhttp3</groupId>
-            <artifactId>okhttp</artifactId>
-            <version>3.14.1</version>
             <scope>test</scope>
         </dependency>
     </dependencies>
@@ -128,13 +128,21 @@
                         <compilerArg>--enable-preview</compilerArg>
                     </compilerArgs>
                 </configuration>
-            </plugin>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-surefire-plugin</artifactId>
-                <configuration>
-                    <argLine>--enable-preview</argLine>
-                </configuration>
+                <executions>
+                    <execution>
+                        <id>default-testCompile</id>
+                        <configuration>
+                            <!--
+                            We use HTTP client from tests, and we do not want it as a module dependency
+                             -->
+                            <compilerArgs>
+                                <compilerArg>--enable-preview</compilerArg>
+                                <compilerArg>--add-modules</compilerArg>
+                                <compilerArg>java.net.http</compilerArg>
+                            </compilerArgs>
+                        </configuration>
+                    </execution>
+                </executions>
             </plugin>
         </plugins>
     </build>

--- a/tests/apps/bookstore/bookstore-se/src/main/java/io/helidon/tests/apps/bookstore/se/Main.java
+++ b/tests/apps/bookstore/bookstore-se/src/main/java/io/helidon/tests/apps/bookstore/se/Main.java
@@ -73,21 +73,8 @@ public final class Main {
      * @return the created {@link WebServer} instance
      */
     static WebServer startServer(boolean ssl, boolean http2, boolean compression) {
-        // load logging configuration
-        LogConfig.configureRuntime();
-
-        // By default this will pick up application.yaml from the classpath
-        Config config = Config.create();
-
-        // Build server config based on params
-        var serverBuilder = WebServer.builder()
-                .addRouting(createRouting(config))
-                .config(config.get("server"))
-                .update(it -> configureJsonSupport(it, config))
-                .update(it -> configureSsl(it, ssl));
-        // .enableCompression(compression);
-
-        configureJsonSupport(serverBuilder, config);
+        WebServerConfig.Builder serverBuilder = WebServerConfig.builder();
+        setupServer(serverBuilder, ssl);
 
         WebServer server = serverBuilder.build();
         server.start();
@@ -96,6 +83,21 @@ public final class Main {
                                    + ", compression=" + compression + "]");
 
         return server;
+    }
+
+    static void setupServer(WebServerConfig.Builder serverBuilder, boolean ssl) {
+        // load logging configuration
+        LogConfig.configureRuntime();
+
+        // By default this will pick up application.yaml from the classpath
+        Config config = Config.create();
+
+        // Build server config based on params
+        serverBuilder
+                .addRouting(createRouting(config))
+                .config(config.get("server"))
+                .update(it -> configureJsonSupport(it, config))
+                .update(it -> configureSsl(it, ssl));
     }
 
     static JsonLibrary getJsonLibrary(Config config) {

--- a/tests/apps/bookstore/bookstore-se/src/main/resources/application.yaml
+++ b/tests/apps/bookstore/bookstore-se/src/main/resources/application.yaml
@@ -18,7 +18,8 @@ app:
   json-library: jsonp
 
 server:
-  port: 8080
+  # Random port
+  port: -1
   host: 0.0.0.0
 
 #  ssl:

--- a/tests/apps/bookstore/bookstore-se/src/main/resources/logging.properties
+++ b/tests/apps/bookstore/bookstore-se/src/main/resources/logging.properties
@@ -31,4 +31,4 @@ java.util.logging.SimpleFormatter.format=%1$tY.%1$tm.%1$td %1$tH:%1$tM:%1$tS %4$
 #io.helidon.config.level=INFO
 #io.helidon.security.level=INFO
 #io.helidon.common.level=INFO
-#io.netty.level=INFO
+#io.helidon.nima.level=ALL

--- a/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/Http2SslTest.java
+++ b/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/Http2SslTest.java
@@ -16,26 +16,25 @@
 
 package io.helidon.tests.apps.bookstore.se;
 
+import io.helidon.common.configurable.ThreadPoolSupplier;
+import io.helidon.nima.testing.junit5.webserver.ServerTest;
+import io.helidon.nima.testing.junit5.webserver.SetUpServer;
+import io.helidon.nima.webserver.WebServer;
+import io.helidon.nima.webserver.WebServerConfig;
+import org.junit.jupiter.api.Test;
+
+import javax.net.ssl.SSLContext;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
 import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.TimeUnit;
 
-import io.helidon.common.configurable.ThreadPoolSupplier;
-
-import io.helidon.nima.webserver.WebServer;
-import okhttp3.OkHttpClient;
-import okhttp3.Protocol;
-import okhttp3.Request;
-import okhttp3.RequestBody;
-import okhttp3.Response;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Disabled;
-import org.junit.jupiter.api.Test;
-
-import static io.helidon.tests.apps.bookstore.se.TestServer.APPLICATION_JSON;
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.jupiter.api.Assertions.fail;
@@ -43,21 +42,25 @@ import static org.junit.jupiter.api.Assertions.fail;
 /**
  * Tests SSL/TLS with HTTP 2 upgrades and compression.
  */
-@Disabled("https://github.com/helidon-io/helidon/issues/7097")
+@ServerTest
 public class Http2SslTest {
 
-    private static WebServer webServer;
-    private static OkHttpClient client;
+    private static HttpClient client;
+    private final URI baseSslUri;
 
-    @BeforeAll
-    public static void startServer() throws Exception {
-        webServer = TestServer.start(true, true, true);
-        client = TestServer.newOkHttpClient(true, true);
+    public Http2SslTest(WebServer server) throws Exception {
+        SSLContext sslContext = TestServer.setupSSLTrust();
+        baseSslUri = URI.create("https://localhost:" + server.port());
+        client = HttpClient.newBuilder()
+                .version(HttpClient.Version.HTTP_2)
+                .sslContext(sslContext)
+                .connectTimeout(Duration.ofSeconds(5))
+                .build();
     }
 
-    @AfterAll
-    public static void stopServer() throws Exception {
-        TestServer.stop(webServer);
+    @SetUpServer
+    static void server(WebServerConfig.Builder server) {
+        Main.setupServer(server, true);
     }
 
     @Test
@@ -72,34 +75,48 @@ public class Http2SslTest {
 
     @Test
     public void testHelloWorldHttp2SslPostFirst() throws Exception {
-        Request.Builder builder = TestServer.newRequestBuilder(webServer, "/books", true);
-        Request postBook = builder.post(
-                RequestBody.create(APPLICATION_JSON, TestServer.getBookAsJson())).build();
-        try (Response postBookRes = client.newCall(postBook).execute()) {
-            assertThat(postBookRes.code(), is(200));
-            assertThat(postBookRes.protocol(), is(Protocol.HTTP_2));
-        }
+        HttpRequest postBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", "gzip")
+                .POST(HttpRequest.BodyPublishers.ofString(TestServer.getBookAsJson()))
+                .build();
 
-        builder = TestServer.newRequestBuilder(webServer, "/books/123456", true);
-        Request deleteBook = builder.delete().build();
-        try (Response deleteBookRes = client.newCall(deleteBook).execute()) {
-            assertThat(deleteBookRes.code(), is(200));
-            assertThat(deleteBookRes.protocol(), is(Protocol.HTTP_2));
-        }
+        var postBookRes = client.send(postBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(postBookRes.statusCode(), is(200));
+        assertThat(postBookRes.version(), is(HttpClient.Version.HTTP_2));
+
+        HttpRequest deleteBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", "gzip")
+                .DELETE()
+                .build();
+
+        var deleteBookRes = client.send(deleteBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(deleteBookRes.statusCode(), is(200));
+        assertThat(deleteBookRes.version(), is(HttpClient.Version.HTTP_2));
+
     }
 
     @Test
     public void testHelloWorldHttp2SslConcurrent() throws Exception {
         ExecutorService executor = ThreadPoolSupplier.create("test-thread-pool").get();
-        Request.Builder builder = TestServer.newRequestBuilder(webServer, "/books", true);
-        Request getBooks = builder.build();
 
-        List<Callable<Response>> tasks = Collections.nCopies(10, () -> client.newCall(getBooks).execute());
+        HttpRequest getBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", "gzip")
+                .GET()
+                .build();
+
+        List<Callable<HttpResponse<String>>> tasks =
+                Collections.nCopies(10, () -> client.send(getBookReq, HttpResponse.BodyHandlers.ofString()));
         executor.invokeAll(tasks).forEach(f -> {
             try {
-                Response r = f.get(1, TimeUnit.SECONDS);
-                assertThat(r.code(), is(200));
-                assertThat(r.protocol(), is(Protocol.HTTP_2));
+                HttpResponse<String> r = f.get(1, TimeUnit.SECONDS);
+                assertThat(r.statusCode(), is(200));
+                assertThat(r.version(), is(HttpClient.Version.HTTP_2));
             } catch (Exception e) {
                 fail(e);
             }
@@ -107,39 +124,61 @@ public class Http2SslTest {
     }
 
     private void testHelloWorld(boolean compression) throws Exception {
-        Request.Builder builder = TestServer.newRequestBuilder(webServer, "/books", true, compression);
+        HttpRequest getBooksReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", compression ? "gzip" : "none")
+                .GET()
+                .build();
 
-        Request getBooks = builder.build();
-        try (Response getBooksRes = client.newCall(getBooks).execute()) {
-            assertThat(getBooksRes.code(), is(200));
-            assertThat(getBooksRes.protocol(), is(Protocol.HTTP_2));
-        }
+        var getBooksRes = client.send(getBooksReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(getBooksRes.statusCode(), is(200));
+        assertThat(getBooksRes.version(), is(HttpClient.Version.HTTP_2));
 
-        Request postBook = builder.post(
-                RequestBody.create(APPLICATION_JSON, TestServer.getBookAsJson())).build();
-        try (Response postBookRes = client.newCall(postBook).execute()) {
-            assertThat(postBookRes.code(), is(200));
-            assertThat(postBookRes.protocol(), is(Protocol.HTTP_2));
-        }
+        HttpRequest postBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", compression ? "gzip" : "none")
+                .POST(HttpRequest.BodyPublishers.ofString(TestServer.getBookAsJson()))
+                .build();
 
-        builder = TestServer.newRequestBuilder(webServer, "/books/123456", true, compression);
-        Request getBook = builder.build();
-        try (Response getBookRes = client.newCall(getBook).execute()) {
-            assertThat(getBookRes.code(), is(200));
-            assertThat(getBookRes.protocol(), is(Protocol.HTTP_2));
-        }
+        var postBookRes = client.send(postBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(postBookRes.statusCode(), is(200));
+        assertThat(postBookRes.version(), is(HttpClient.Version.HTTP_2));
 
-        Request deleteBook = builder.delete().build();
-        try (Response deleteBookRes = client.newCall(deleteBook).execute()) {
-            assertThat(deleteBookRes.code(), is(200));
-            assertThat(deleteBookRes.protocol(), is(Protocol.HTTP_2));
+        HttpRequest getBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", compression ? "gzip" : "none")
+                .GET()
+                .build();
 
-        }
+        var getBookRes = client.send(getBookReq, HttpResponse.BodyHandlers.ofString());
 
-        Request getNoBook = builder.build();
-        try (Response getNoBookRes = client.newCall(getNoBook).execute()) {
-            assertThat(getNoBookRes.code(), is(404));
-            assertThat(getNoBookRes.protocol(), is(Protocol.HTTP_2));
-        }
+        assertThat(getBookRes.statusCode(), is(200));
+        assertThat(getBookRes.version(), is(HttpClient.Version.HTTP_2));
+
+        HttpRequest deleteBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", compression ? "gzip" : "none")
+                .DELETE()
+                .build();
+
+        var deleteBookRes = client.send(deleteBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(deleteBookRes.statusCode(), is(200));
+        assertThat(deleteBookRes.version(), is(HttpClient.Version.HTTP_2));
+
+        HttpRequest getNoBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .version(HttpClient.Version.HTTP_2)
+                .setHeader("Accept-Encoding", compression ? "gzip" : "none")
+                .GET()
+                .build();
+
+        var getNoBookRes = client.send(getNoBookReq, HttpResponse.BodyHandlers.ofString());
+
+        assertThat(getNoBookRes.statusCode(), is(404));
+        assertThat(getNoBookRes.version(), is(HttpClient.Version.HTTP_2));
     }
 }

--- a/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/MainTest.java
+++ b/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/MainTest.java
@@ -16,77 +16,98 @@
 
 package io.helidon.tests.apps.bookstore.se;
 
-
+import io.helidon.nima.testing.junit5.webserver.ServerTest;
+import io.helidon.nima.testing.junit5.webserver.SetUpServer;
 import io.helidon.nima.webserver.WebServer;
+import io.helidon.nima.webserver.WebServerConfig;
 import jakarta.json.Json;
 import jakarta.json.JsonObject;
 import jakarta.json.JsonReader;
-import okhttp3.OkHttpClient;
-import okhttp3.Request;
-import okhttp3.RequestBody;
-import okhttp3.Response;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
 
-import static io.helidon.tests.apps.bookstore.se.TestServer.APPLICATION_JSON;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.CoreMatchers.notNullValue;
 import static org.hamcrest.MatcherAssert.assertThat;
 
+@ServerTest
 public class MainTest {
 
-    private static WebServer webServer;
-    private static OkHttpClient client;
+    private static HttpClient client;
+    private final URI baseUri;
 
-    @BeforeAll
-    public static void startServer() throws Exception {
-        webServer = TestServer.start(false, false, false);
-        client = TestServer.newOkHttpClient(false, false);
+    public MainTest(WebServer server) throws Exception {
+        baseUri = URI.create("http://localhost:" + server.port());
+        client = HttpClient.newBuilder()
+                .version(HttpClient.Version.HTTP_1_1)
+                .connectTimeout(Duration.ofSeconds(5))
+                .build();
     }
 
-    @AfterAll
-    public static void stopServer() throws Exception {
-        TestServer.stop(webServer);
+    @SetUpServer
+    static void server(WebServerConfig.Builder server) {
+        Main.setupServer(server, false);
     }
 
     @Test
     public void testHelloWorld() throws Exception {
-        Request.Builder builder = TestServer.newRequestBuilder(webServer, "/books", false);
+        HttpRequest getBooksReq = HttpRequest.newBuilder()
+                .uri(baseUri.resolve("/books"))
+                .GET()
+                .build();
 
-        Request getBooks = builder.build();
-        try (Response getBooksRes = client.newCall(getBooks).execute()) {
-            assertThat(getBooksRes.code(), is(200));
-            assertThat(getBooksRes.header("content-length"), notNullValue());
-            String body = getBooksRes.body().string();
-            assertThat(body, is("[]"));
-        }
+        var getBooksRes = client.send(getBooksReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(getBooksRes.statusCode(), is(200));
+        assertThat(getBooksRes.version(), is(HttpClient.Version.HTTP_1_1));
+        assertThat(getBooksRes.headers().firstValue("content-length").orElse(null), notNullValue());
+        assertThat(getBooksRes.body(), is("[]"));
 
-        Request postBook = builder.post(
-                RequestBody.create(APPLICATION_JSON, TestServer.getBookAsJson())).build();
-        try (Response postBookRes = client.newCall(postBook).execute()) {
-            assertThat(postBookRes.code(), is(200));
-        }
+        HttpRequest postBookReq = HttpRequest.newBuilder()
+                .uri(baseUri.resolve("/books"))
+                .POST(HttpRequest.BodyPublishers.ofString(TestServer.getBookAsJson()))
+                .build();
 
-        builder = TestServer.newRequestBuilder(webServer, "/books/123456", false);
-        Request getBook = builder.build();
-        try (Response getBookRes = client.newCall(getBook).execute()) {
-            assertThat(getBookRes.code(), is(200));
-            assertThat(getBookRes.header("content-length"), notNullValue());
-            JsonReader jsonReader = Json.createReader(getBookRes.body().byteStream());
-            JsonObject jsonObject = jsonReader.readObject();
-            assertThat("Checking if correct ISBN", jsonObject.getString("isbn"),
-                    is("123456"));
-        }
+        var postBookRes = client.send(postBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(postBookRes.statusCode(), is(200));
+        assertThat(postBookRes.version(), is(HttpClient.Version.HTTP_1_1));
 
-        Request deleteBook = builder.delete().build();
-        try (Response deleteBookRes = client.newCall(deleteBook).execute()) {
-            assertThat(deleteBookRes.code(), is(200));
-        }
+        HttpRequest getBookReq = HttpRequest.newBuilder()
+                .uri(baseUri.resolve("/books/123456"))
+                .GET()
+                .build();
 
-        Request getNoBook = builder.build();
-        try (Response getNoBookRes = client.newCall(getNoBook).execute()) {
-            assertThat(getNoBookRes.code(), is(404));
-        }
+        var getBookRes = client.send(getBookReq, HttpResponse.BodyHandlers.ofInputStream());
+
+        assertThat(getBookRes.statusCode(), is(200));
+        assertThat(getBookRes.version(), is(HttpClient.Version.HTTP_1_1));
+        assertThat(getBooksRes.headers().firstValue("content-length").orElse(null), notNullValue());
+        JsonReader jsonReader = Json.createReader(getBookRes.body());
+        JsonObject jsonObject = jsonReader.readObject();
+        assertThat("Checking if correct ISBN", jsonObject.getString("isbn"), is("123456"));
+
+
+        HttpRequest deleteBookReq = HttpRequest.newBuilder()
+                .uri(baseUri.resolve("/books/123456"))
+                .DELETE()
+                .build();
+
+        var deleteBookRes = client.send(deleteBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(deleteBookRes.statusCode(), is(200));
+        assertThat(deleteBookRes.version(), is(HttpClient.Version.HTTP_1_1));
+
+        HttpRequest getNoBookReq = HttpRequest.newBuilder()
+                .uri(baseUri.resolve("/books/123456"))
+                .GET()
+                .build();
+
+        var getNoBookRes = client.send(getNoBookReq, HttpResponse.BodyHandlers.ofString());
+
+        assertThat(getNoBookRes.statusCode(), is(404));
+        assertThat(getNoBookRes.version(), is(HttpClient.Version.HTTP_1_1));
     }
 }

--- a/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/SslTest.java
+++ b/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/SslTest.java
@@ -16,68 +16,94 @@
 
 package io.helidon.tests.apps.bookstore.se;
 
+import io.helidon.nima.testing.junit5.webserver.ServerTest;
+import io.helidon.nima.testing.junit5.webserver.SetUpServer;
 import io.helidon.nima.webserver.WebServer;
 
-import okhttp3.OkHttpClient;
-import okhttp3.Request;
-import okhttp3.RequestBody;
-import okhttp3.Response;
-import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.BeforeAll;
+import io.helidon.nima.webserver.WebServerConfig;
 import org.junit.jupiter.api.Test;
 
-import static io.helidon.tests.apps.bookstore.se.TestServer.APPLICATION_JSON;
+import javax.net.ssl.SSLContext;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 
 /**
  * Tests SSL/TLS with HTTP 1.1.
  */
+@ServerTest
 public class SslTest {
 
-    private static WebServer webServer;
-    private static OkHttpClient client;
+    private static HttpClient client;
+    private final URI baseSslUri;
 
-    @BeforeAll
-    public static void startServer() throws Exception {
-        webServer = TestServer.start(true, false, false);
-        client = TestServer.newOkHttpClient(true, false);
+    public SslTest(WebServer server) throws Exception {
+        SSLContext sslContext = TestServer.setupSSLTrust();
+        baseSslUri = URI.create("https://localhost:" + server.port());
+        client = HttpClient.newBuilder()
+                .sslContext(sslContext)
+                .version(HttpClient.Version.HTTP_1_1)
+                .connectTimeout(Duration.ofSeconds(5))
+                .build();
     }
 
-    @AfterAll
-    public static void stopServer() throws Exception {
-        TestServer.stop(webServer);
+    @SetUpServer
+    static void server(WebServerConfig.Builder server) {
+        Main.setupServer(server, true);
     }
 
     @Test
     public void testHelloWorldSsl() throws Exception {
-        Request.Builder builder = TestServer.newRequestBuilder(webServer, "/books", true);
+        HttpRequest getBooksReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .GET()
+                .build();
 
-        Request getBooks = builder.build();
-        try (Response getBooksRes = client.newCall(getBooks).execute()) {
-            assertThat(getBooksRes.code(), is(200));
-        }
+        var getBooksRes = client.send(getBooksReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(getBooksRes.statusCode(), is(200));
+        assertThat(getBooksRes.version(), is(HttpClient.Version.HTTP_1_1));
 
-        Request postBook = builder.post(
-                RequestBody.create(APPLICATION_JSON, TestServer.getBookAsJson())).build();
-        try (Response postBookRes = client.newCall(postBook).execute()) {
-            assertThat(postBookRes.code(), is(200));
-        }
+        HttpRequest postBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books"))
+                .POST(HttpRequest.BodyPublishers.ofString(TestServer.getBookAsJson()))
+                .build();
 
-        builder = TestServer.newRequestBuilder(webServer, "/books/123456", true);
-        Request getBook = builder.build();
-        try (Response getBookRes = client.newCall(getBook).execute()) {
-            assertThat(getBookRes.code(), is(200));
-        }
+        var postBookRes = client.send(postBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(postBookRes.statusCode(), is(200));
+        assertThat(postBookRes.version(), is(HttpClient.Version.HTTP_1_1));
 
-        Request deleteBook = builder.delete().build();
-        try (Response deleteBookRes = client.newCall(deleteBook).execute()) {
-            assertThat(deleteBookRes.code(), is(200));
-        }
+        HttpRequest getBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .GET()
+                .build();
 
-        Request getNoBook = builder.build();
-        try (Response getNoBookRes = client.newCall(getNoBook).execute()) {
-            assertThat(getNoBookRes.code(), is(404));
-        }
+        var getBookRes = client.send(getBookReq, HttpResponse.BodyHandlers.ofString());
+
+        assertThat(getBookRes.statusCode(), is(200));
+        assertThat(getBookRes.version(), is(HttpClient.Version.HTTP_1_1));
+
+        HttpRequest deleteBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .DELETE()
+                .build();
+
+        var deleteBookRes = client.send(deleteBookReq, HttpResponse.BodyHandlers.ofString());
+        assertThat(deleteBookRes.statusCode(), is(200));
+        assertThat(deleteBookRes.version(), is(HttpClient.Version.HTTP_1_1));
+
+        HttpRequest getNoBookReq = HttpRequest.newBuilder()
+                .uri(baseSslUri.resolve("/books/123456"))
+                .GET()
+                .build();
+
+        var getNoBookRes = client.send(getNoBookReq, HttpResponse.BodyHandlers.ofString());
+
+        assertThat(getNoBookRes.statusCode(), is(404));
+        assertThat(getNoBookRes.version(), is(HttpClient.Version.HTTP_1_1));
     }
 }

--- a/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/TestServer.java
+++ b/tests/apps/bookstore/bookstore-se/src/test/java/io/helidon/tests/apps/bookstore/se/TestServer.java
@@ -16,37 +16,45 @@
 
 package io.helidon.tests.apps.bookstore.se;
 
-import javax.net.ssl.SSLContext;
-import javax.net.ssl.TrustManager;
-import javax.net.ssl.X509TrustManager;
+import io.helidon.nima.testing.junit5.webserver.SetUpServer;
+import io.helidon.nima.webserver.WebServerConfig;
+
+import javax.net.ssl.*;
 import java.io.BufferedReader;
-import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.net.URL;
+import java.net.Socket;
 import java.security.SecureRandom;
+import java.security.cert.CertificateException;
 import java.security.cert.X509Certificate;
-import java.util.Arrays;
-import java.util.List;
 import java.util.stream.Collectors;
-
-import io.helidon.nima.webserver.WebServer;
-import okhttp3.Interceptor;
-import okhttp3.MediaType;
-import okhttp3.OkHttpClient;
-import okhttp3.Protocol;
-import okhttp3.Request;
-import okhttp3.Response;
-import org.junit.jupiter.api.Assertions;
 
 /**
  * Utility class to start the Helidon server with options.
  */
 class TestServer {
 
-    static final MediaType APPLICATION_JSON = MediaType.parse("application/json");
+    private static final X509TrustManager NOOP_TRUST_MANAGER = new X509ExtendedTrustManager() {
+        @Override
+        public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+                throws CertificateException {
+        }
 
-    private static X509TrustManager TRUST_MANAGER = new X509TrustManager() {
+        @Override
+        public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+                throws CertificateException {
+        }
+
+        @Override
+        public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+                throws CertificateException {
+        }
+
+        @Override
+        public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine) throws
+                CertificateException {
+        }
+
         public X509Certificate[] getAcceptedIssuers() {
             return new X509Certificate[0];
         }
@@ -57,28 +65,6 @@ class TestServer {
         public void checkServerTrusted(X509Certificate[] certs, String authType) {
         }
     };
-
-    static WebServer start(boolean ssl, boolean http2, boolean compression) throws Exception {
-        WebServer webServer = Main.startServer(ssl, http2, compression);
-
-        long timeout = 2000; // 2 seconds should be enough to start the server
-        long now = System.currentTimeMillis();
-
-        while (!webServer.isRunning()) {
-            Thread.sleep(100);
-            if ((System.currentTimeMillis() - now) > timeout) {
-                Assertions.fail("Failed to start webserver");
-            }
-        }
-
-        return webServer;
-    }
-
-    static void stop(WebServer webServer) throws Exception {
-        if (webServer != null) {
-            webServer.stop();
-        }
-    }
 
     static String getBookAsJson() throws Exception {
         try (InputStream is = TestServer.class.getClassLoader().getResourceAsStream("book.json")) {
@@ -92,53 +78,7 @@ class TestServer {
 
     static SSLContext setupSSLTrust() throws Exception {
         SSLContext sslContext = SSLContext.getInstance("TLS");
-        sslContext.init(null, new TrustManager[] { TRUST_MANAGER }, new SecureRandom());
+        sslContext.init(null, new TrustManager[] {NOOP_TRUST_MANAGER}, new SecureRandom());
         return sslContext;
-    }
-
-    static Request.Builder newRequestBuilder(WebServer webServer, String path, boolean ssl) throws Exception {
-        return newRequestBuilder(webServer, path, ssl, false);
-    }
-
-    static Request.Builder newRequestBuilder(WebServer webServer, String path, boolean ssl, boolean compression)
-            throws Exception {
-        URL url = new URL((ssl ? "https" : "http") + "://localhost:" + webServer.port() + path);
-        Request.Builder builder = new Request.Builder().url(url);
-        builder.addHeader("Accept-Encoding", compression ? "gzip" : "none");
-        return builder;
-    }
-
-    static class LoggingInterceptor implements Interceptor {
-        @Override
-        public Response intercept(Interceptor.Chain chain) throws IOException {
-            Request request = chain.request();
-
-            long t1 = System.nanoTime();
-            System.out.println(String.format("Sending request %s %s on %s%n%s",
-                    request.method(), request.url(), chain.connection(), request.headers()));
-
-            Response response = chain.proceed(request);
-
-            long t2 = System.nanoTime();
-            System.out.println(String.format("Received response %d for %s in %.1fms%nProtocol is %s%n%s",
-                    response.code(), response.request().url(), (t2 - t1) / 1e6d, response.protocol(),
-                    response.headers()));
-
-            return response;
-        }
-    }
-
-    static OkHttpClient newOkHttpClient(boolean ssl, boolean http2) throws Exception {
-        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
-                .addNetworkInterceptor(new LoggingInterceptor())
-                .retryOnConnectionFailure(false)
-                .protocols(http2 ? Arrays.asList(Protocol.HTTP_2, Protocol.HTTP_1_1)
-                                 : List.of(Protocol.HTTP_1_1));
-        if (ssl) {
-            SSLContext sslContext = setupSSLTrust();
-            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), TRUST_MANAGER);
-            clientBuilder.hostnameVerifier((host, session) -> host.equals("localhost"));
-        }
-        return clientBuilder.build();
     }
 }

